# Ejercicios

 1. Si en una determinada calculadora trabaja solo con 4 cifras significativas y con truncamiento, ¿qué devolverá como resultado de $66.666 \times 33.333$? ¿y si trabajase con redondeo?

In [1]:
# Nota: la librería sigfig (https://pypi.org/project/sigfig/) permite redondear resultados a X cifras significativas. Para hacerlo, primero instala la librería ejecutando
%pip install -q sigfig

#Luego, por ejemplo
import sigfig
 # 3.1

Note: you may need to restart the kernel to use updated packages.


In [10]:
sigfig.round(3.14159, sigfigs=5) 

3.1416

In [17]:
import sigfig

a = 66.666
b = 33.333
producto = a * b

print(producto)

#producto_redondeado = sigfig.round(producto, figs=4)

producto_redondeado = f"{producto:.4g}"

print(producto_redondeado)


2222.1777779999998
2222


2. Indica si las siguientes afirmaciones sobre la representación de coma
  flotante son verdaderas o falsas:
 - En cada operación que se realiza se añade un número pequeño numero
        aleatorio imprevisible de orden de la precisión de máquina $\epsilon$. Por ejemplo,
        $1 + 1$ dará $2 \pm \epsilon$ , y $0 \times 7$ dará $\pm \epsilon$.
 - Las operaciones con enteros son intrínsecamente más precisas que las
        operaciones con reales.
 - En la representación de coma flotante, hay una representación para el 0
        positivo y otra para el 0 negativo.
 - Sean $m$ y $M$ el menor y el mayor número representable en coma flotante
        con 64bits. Se cumple que tanto $m/M$ como $M/m$ resultan en un
        desbordamiento.
 - El error de redondeo es siempre menor o igual que el de truncamiento.
 - La principal ventaja de la representación de coma flotante frente a la
        notación habitual es que el *gap* entre *floats* se mantiene
        aproximadamente constante, sea cual sea el número representado.

3. Convierte los siguientes números a binario:
   - $69$
   - $255$
   - $34.5$
   - $3.1416$
   - $5.\widehat{3}$

In [17]:
def binario(n: float) -> str:
    cadena = ''

    while n > 1:
        modulo = str(n % 2)
        cadena += modulo
        n //= 2

    if n == 1:
        cadena += '1'

    if n == 0:
        return '0'

    return ''.join(reversed(cadena))

binario(69)
binario(255)

'11111111'

5. Convierte las siguientes representaciones en coma flotante de 64bits a
  números decimales.
 - $$\quad 0 \quad 10000000111 \quad 11100000000000000000000000000000000000000000$$ 
 - $$\quad 1 \quad 10000001001 \quad 10000000000000000000000000000000000000000000$$ 
 - $$\quad 1 \quad 11111110000 \quad 00001000000000000000000000000000000000000000$$

6. Obtén la representación de coma flotante de 64bits de los siguientes números:
   - $100$
   - $74.1$
   - $0.00095$


7. Una solución que se utiliza frecuentemente para el problema de la
  *cancelación catastrófica* es el de multiplicar y dividir por el conjugado de 
  la diferencia que se está calculando. Comprueba que esta
  técnica funciona resolviendo la siguiente ecuación $x^2 - 20x + 1 = 0$ con
  solo dos cifras significativas. Compara el error relativo cometido usando y
  sin usar la técnica anterior. 
  
    *Pista: la cancelación catastrófica aparece al aplicar la fórmula de resolución de ecuaciones de segundo grado, en el numerador.*

In [18]:
# Nota: la librería sigfig (https://pypi.org/project/sigfig/) permite redondear resultados a X cifras significativas. Para hacerlo, primero instala la librería ejecutando
%pip install -q sigfig

#Luego, por ejemplo
import sigfig
sigfig.round(3.14159, sigfigs=2)  # 3.1

Note: you may need to restart the kernel to use updated packages.


3.1

8. En precisión simple (32bits) el signo ocupa 1 bit, el exponente 8 y la
mantisa 23. El sesgo del exponente es 127. Obtén el mayor y el menor número
representable.


9. Usando Python, determina el error de máquina escribiendo un programa que
  calcule sucesivamente la expresión $1+2^{1/i}$ para $i=1,2,\dots$ mientras que
  sea mayor que $1$.



10. Considera la sucesión de Fibonacci:

$$
\left\{\begin{array}{l}
x_{0}=1, x_{1}=1 \\
x_{n+1}=x_{n}+x_{n-1}, n \in \mathbb{N} .
\end{array}\right.
$$

Usando Python, estima el siguiente límite:

$$\lim_{n \to \infty} \frac{x_n}{x_{n-1}}$$

y grafica el error relativo frente a $n$, sabiendo que la expresión exacta del
límite anterior es el número áureo $\phi \simeq 1.61803$.


11. En Python, escribe dos funciones ```dec_to_bin()``` y ```bin_to_dec()``` para
  convertir números **enteros** de decimal a binario y viceversa. Almacena los
  números binarios en una variables ```string```. Escribe un test que llame a
  ambas funciones de forma anidada comprobando que el *output* es el mismo que
  el *input*, para valores $i = 1, 2, \dots, 10000$.

12. Extiende las funciones de apartado anterior para que puedan trabajar con
  números reales.

13. Programa una función que calcule el error absoluto y relativo de aproximar una función $f(x)$ mediante su serie de Taylor en $x=0$. Los argumentos de dicha función deben permitir seleccionar al usuario entre la función que desea aproximar de entre [$e^x$, $\cos x$, $\sin x$], el número de términos que quiere usar ($n$) y el valor donde evaluar dicha aproximación ($x$). Utiliza las expresiones del desarrollo de Taylor que conoces de [$e^x$, $\cos x$, $\sin x$].

14. Utiliza la función del ejercicio anterior para mostrar en una
gráfica cómo evoluciona el error relativo frente a $n$ y frente a $x$. Comenta
los resultados.


15. En la siguiente imagen se muestra el mecanismo que
mueve un pistón. 

<center>
<div>
<img src="https://drive.upm.es/index.php/apps/files_sharing/ajax/publicpreview.php?x=1920&y=457&a=true&file=piston.png&t=yZH10VCe9BAfnbn&scalingup=0" width="300"/>
</div>
</center>

Se han medido el tamaño del brazo menor: $R = 12cm$ con un error de $\pm3\%$. Se
sabe que la distancia entre la articulación $A$ y el pistón $C$ viene dada por
las siguiente expresión: 
$$x = R \left( \cos \theta + \sqrt{2.5^2 - \sin ^2 \theta} \right)$$
¿entre qué valores variará $x$ para $\theta = \pi /2 \pm 0.1$?


16. Se quieren encontrar las raíces de la ecuación $x^2 - 4998x + 1 = 0$. Supón aritmética de 8 dígitos. Obtén el resultado utilizando la fórmula habitual de ecuaciones de segundo grado. Piensa en alguna modificación que permita conservar más cifras significativas.

17. La sucesión definida por recurrencia mediante
$$
x_{n+1} = 2^{n+1} \left( \sqrt{1 + 2^{-n} \cdot x_n} - 1 \right) \qquad n \ge 0
$$
converge a $\ln(1 + x_0)$ para cualquier valor $x_0 > -1$.
 - Reescribir la expresión anterior para evitar la pérdida de cifras significativas.
 - Incluso con la expresión del apartado anterior los resultados numéricos darán $x_{n+1} = x_{n}$ a partir de un cierto valor de $n$. ¿Por qué? Estimar este valor de $n$ para $x_0 = 2$ y aritmética de doble precisión.

18. Como ya se ha visto, en precisión doble se reserva 1 bit para el signo $s$, $11$
  para el exponente $e$ y $52$ la mantisa $m$. Dada la tupla $(s,e,m)$, el
  número representado puede obtenerse entonces con la siguiente fórmula:
  $$x_F = \left( -1 \right)^{s} \cdot 1.f \cdot 2^{e-b}$$
  siendo $b$ el *sesgo*, que se obtiene dividiendo por dos el mayor número representable con
  los bits del exponente, y restándole al resultado 1 (en precisión doble,
  $(11111111111)_2 = 2048 \rightarrow b = 2048/2 - 1 = 1023$). Supón ahora que
  en lugar de $11$ bits para el exponente se reservan $n$, quedando $63-n$ para
  la mantisa. Usando Python, calcula para $n  = 5, 6, \dots, 15$ el mayor y el menor número
  según la representación de coma flotante descrita.

19. Usando Python, escribe una clase ```MyBin``` que represente los enteros en
    base 2. En particular, sobrecarga el operador $+$ para que sea capaz de
    sumar números binarios (no deben convertirse a decimal). Programa también
    los siguientes test para comprobar el funcionamiento del operador:
    
         - ```MyBin("11111") + MyBin("1") #Output: MyBin("100000")```
         - ```MyBin("11111") + MyBin("1010100") #Output: MyBin("1110011")```
         - ```MyBin("110") + MyBin("101") #Output: MyBin("1011")```


20. A continuación se dan varias funciones que presentan algún
    comportamiento característico en $x=0$. Calcular aproximaciones de la
    derivada en $x=0$ por un esquema de diferencias finitas de orden 2
    centrado, y de orden 1 adelantado y atrasado. Toma $h=0.005$ (constante). Interpreta los
    resultados, calculando:
 - $f_1(x) = 0.2x^3+2x-4$ (alta linealidad)
 - $f_2(x)-\ln(x+0.01)$ (gradiente negativo grande)
 - $f_3(x) = 15/\left(1+15x^2\right)$ (máximo con cambio brusco de gradiente)
 - $f_4(x) = x \sin(1/(x+0.005))$ (máximos y mínimos sucesivamente próximos)


21. Las siguientes expresiónes corresponden con las fórmulas para aproximar $f'(x)$ y $f''(x)$ por diferencias finitas atrasadas de primer orden:
$$
f^{\prime}(x) \approx \frac{f(x)-f(x-h)}{h}
$$

$$
f^{\prime \prime}(x) \approx \frac{f(x-2 h)-2 f(x-h)+f(x)}{h^{2}}
$$
 
Obtén la expresión para $f'''(x)$ derivando en la
  expresión de $f''(x)$.

22. Deduce a partir de desarrollos de Taylor la expresión que permite aproximar
  la derivada tercera $f'''(x)$ usando diferencias finitas adelantadas de
  segundo orden.

23. Deduce a partir de desarrollos de Taylor la expresión que permite aproximar
  la derivada cuarta $f^{(4)}(x)$ usando diferencias finitas centradas de
  segundo orden.

24. Utilice aproximaciones de diferencias finitas para:
  
  -  Calcular $f(2.36)$ y $f'(2.36)$ a partir de los siguientes datos (usar DF de orden 2):

<center>

| $x$  | $f(x)$  |
|------|---------|
| 2.36 | 0.85866 |
| 2.37 | 0.86289 |
| 2.38 | 0.86710 |
| 2.39 | 0.87129 |

</center>

- Estimar $f'(1)$ y $f''(1)$ de los siguientes datos (usar DF de orden 2). Presta atención al espaciamiento entre nodos:

<center>

| $x$  | $f(x)$  |
|------|---------|
| 0.97 | 0.85040 |
| 1.00 | 0.84147 |
| 1.05 | 0.82612 |

</center>

- Dados los siguientes datos, estimar $f''(1)$ de la forma más precisa que
  puedas:

<center>

| $x$  | $f(x)$   |
|------|----------|
| 0.84 | 0.431711 |
| 0.92 | 0.398519 |
| 1.00 | 0.367879 |
| 1.08 | 0.339596 |
| 1.16 | 0.313486 |
</center>

- Dados los siguientes datos, estimar $f'(0.2)$ de la forma más precisa que
  puedas:

<center>

| $x$ | $f(x)$   |
|-----|----------|
| 0.0 | 0.000000 |
| 0.1 | 0.078348 |
| 0.2 | 0.138910 |
| 0.3 | 0.192916 |
| 0.4 | 0.244981 |
</center>

25. Utilizando cinco cifras significativas en los cálculos, determina $d(\sin x)/dx$ en $x = 0.8$ a partir de (a) la aproximación de orden uno por diferencias finitas adelantadas y (b) la aproximación por diferencias finitas centradas de orden dos. En cada caso, usa la solución analítica para encontrar el paso $h$ que dé el resultado más exacto (se requiere experimentación).

In [ ]:
# Nota: la librería sigfig (https://pypi.org/project/sigfig/) permite redondear resultados a X cifras significativas. Para hacerlo, primero instala la librería ejecutando
%pip install -q sigfig

#Luego, por ejemplo
import sigfig
sigfig.round(3.14159, sigfigs=2)  # 3.1

26. La biela $AB$ de longitud $R = 90$ mm gira a una velocidad angular constante
de $d \theta / dt = 5000$ rev/min. 

<center>
<div>
<img src="https://drive.upm.es/index.php/apps/files_sharing/ajax/publicpreview.php?x=1920&y=457&a=true&file=piston.png&t=yZH10VCe9BAfnbn&scalingup=0" width="240"/>
</div>
</center>

Se puede demostrar que la posición del pistón $C$ varía con el ángulo $\theta$
como: 

$$ x=R\left(\cos\theta+\sqrt{2.5^{2}-\sin ^{2} \theta}\right) $$ 

Escriba un programa que grafique la aceleración del pistón en $\theta = 0^{\circ}, 5^{\circ},
10^{\circ}, \dots , 180^{\circ}$. Utilice derivación numérica para calcular la aceleración.

27. Las estaciones radar $A$ y $B$, separadas una distancia $a = 500$ m, siguen a
  un avión $C$ guardando los ángulos $\alpha$ y $\beta$ en intervalos de un
  segundo. 
  
<center>
<div>
<img src="https://drive.upm.es/index.php/apps/files_sharing/ajax/publicpreview.php?x=1920&y=457&a=true&file=avion.png&t=UUIC4S71Gj1OWx3&scalingup=0" width="240"/>
</div>
</center>
  
  Si tres lecturas sucesivas fueron:

<center>

| $t$ (s) | $\alpha$        | $\beta$         |
|---------|-----------------|-----------------|
| 9       | $54.80^{\circ}$ | $65.59^{\circ}$ |
| 10      | $54.06^{\circ}$ | $64.59^{\circ}$ |
| 11      | $53.34^{\circ}$ | $63.62^{\circ}$ |

</center>

calcula la velocidad $v$ del avión y el ángulo de ascenso $\gamma$ en $t = 10$s. Puede demostrarse que las coordenadas del avión vienen dadas por:

$$
x=a \frac{\tan \beta}{\tan \beta-\tan \alpha} \quad y=a \frac{\tan \alpha \tan \beta}{\tan \beta-\tan \alpha}
$$

28. Una determinado sistema de medida está midiendo una magnitud $f(t)$ que depende del tiempo $t \in \left[ 0, 2 \pi \right]$. Se sabe que dicha magnitud se comporta como $f(t) = \cos t$. Sin embargo, debido a la sensibilidad del sistema de medida, en verdad la magnitud está siendo almacenada como $f_{\epsilon} (t) = \cos t + \epsilon \sin \omega t$, cumpliéndose siempre que $\epsilon \ll 1$. 
      - Dibuja $f(t)$ y $f_{\epsilon} (t)$ en la misma gráfica, para $t \in \left[ 0, 2 \pi \right]$. Usa $\epsilon = 0.01$ y $\omega = 100$. Observa que ambas funciones son muy similares.
      - Utiliza un esquema de diferencias finitas adelantadas de primer orden para calcular la derivada $f_{\epsilon}^{\prime} (t)$ y grafícala frente a la solución analítica de $f^{\prime} (t)$. Analiza el resultado
      - Piensa en formas de solucionar el problema identificado anteriormente.

29. Escriba una función ```finite_dif(f: Callable, min_val: float, max_val: float, num_div: int)```, donde ```f``` es una instancia de una función a derivar, ```min_val``` y ```max_val``` son los extremos del intervalo considerado (```a``` < ```b```), ```num_div``` es el número de puntos en los que se divide el intervalo anterior (```num_div``` debe ser mayor que $10$), y por último ```option``` toma uno de los siguientes valores: ```"forward"```, ```"backward"``` o ```"central"```. Esta función debe devolver una tupla ```(df, x)```, siendo ```df``` un *array* que contenga la derivada de la función según el método seleccionado (asume $\mathcal{O} \left( h^2 \right)$), y ```x``` los nodos donde es calculada (observa que ```x``` difiere de la división del intervalo original, ya que en función del esquema seleccionado, se perderá alguno de los nodos de los extremos). Utiliza el siguiente test para comprobar tu implementación:
 
```python
import numpy as np
x = np.linspace(0, 2 * np.pi, 100)
f = lambda x: np.sin(x)
df_f, x_f = finite_dif(f, 0, 2 * np.pi, 10, 'forward')
df_b, x_b = finite_dif(f, 0, 2 * np.pi, 10, 'backward')
df_c, x_c = finite_dif(f, 0, 2 * np.pi, 10, 'central')
plt.figure(figsize = (12, 8))
plt.plot(x, np.cos(x), label = 'analytic')
plt.plot(x_f, df_f, label = 'forward')
plt.plot(x_b, df_b, label = 'backward')
plt.plot(x_c, df_c, label = 'central')
plt.legend()
plt.show()
```

30. Para calcular la derivada de una función altamente oscilante (implícita o explícita), un método que puede usarse es realizar un suavizado previo de los valores de la función. Escribe una función ```finite_dif_smoothing(x: np.ndarray, y: np.ndarray, n: int)```, donde ```x``` e ```y``` son *arrays* de la misma longitud que contienen los valores $x$ e $y$ de una determinada función sobre la que se calculará la derivada, y ```n``` es estrictamente positivo. El método debe comenzar calculando un nuevo *array* de valores ```y``` suavizados, donde ```y_smooth[i] = np.mean[i-n:i+n]```. La salida de la función es una tupla ```(dy, x)```, siendo ```dy``` un *array* que contenga la derivada de la función sobre el vector ```y_smooth``` (utiliza diferencias finitas centradas de orden dos), y ```x``` los nodos correspondientes. Supongamos que los datos contenidos en ```x``` están en orden ascendente sin entradas duplicadas. Sin embargo, es posible que los elementos de ```x``` no estén espaciados uniformemente. Ten en cuenta que la salida ```dy``` tendrá menos puntos que ```y```. 

31. Investiga formas de general mallados no uniformes. En particular, piensa un método que permita generar una malla $1D$ en un dominio de longitud $L$ con:
    - Concentración de nodos en ambos extremos del dominio.
    - Concentración en el centro del dominio
    - Concentración solo en uno de los extremos del dominio, hasta una cierta distancia $l \lesssim L$, y constante después.
   
   En todos los casos, establece algún parámetro para controlar la concentración.

32. La ecuación de segundo grado $x^2 - 2x + 1 = 0$ tiene la raíz doble $x = 1$, sin embargo la ecuación $x^2 - 2x + 0.99 = 0$ tiene dos raíces reales $x = 1.1$ y $x = 0.9$, y la ecuación $x^2 - 2x + 1.01 = 0$ tiene dos raíces complejas $x = 1 \pm 0.1i$. Estudia el condicionamiento de $x^2 - 2x + \lambda = 0$ para explicar lo que ocurre. 

33. Usando Python, estudia el condicionamiento de la función $f(x) = \tan x$ (puedes usar la librería ```sympy``` si lo deseas).

34. Escribe en Python una función ```conditioning(f: Callable, df:Callable, x: float)``` para calcular el número de condición al evaluar una función $f(x)$ en un punto $x$. ```f``` y ```df``` son funciones que representan $f(x)$ y $f'(x)$ respectivamente, y ```x``` es el punto donde se desea evaluar el condicionamiento. Utiliza la función anterior para graficar el número de condición de $f(x) = 3(x-1)^5+7(x-1)^9$ para $x>1$. Interpreta el resultado. Comprueba para alguno de los valores donde la función está mal condicionada que efectivamente una pequeña variación de los datos de entrada resulta en una gran variación de los datos de salida.

35. El sistema lineal $Ax=b$ está definido por:
$$
A = \begin{pmatrix}
10 & 7 & 8  & 7 \\
7  & 5 & 6  & 5 \\
8  & 6 & 10 & 9 \\
7  & 5 & 9  & 10
\end{pmatrix} 
\qquad
b = \begin{pmatrix}
32 \\
23 \\
33 \\
31 
\end{pmatrix} 
$$
  * Usando Python, encuentra la solución exacta del problema
  * Aplica la siguiente variación al vector $b$ y resuelve de nuevo el sistema:
$$
\Delta b = \begin{pmatrix}
0.1 \\
-0.1 \\
0.1 \\
-0.1
\end{pmatrix} 
$$
  * Calcula el número de condición $\kappa_2$ e interprétalo
  * Supón ahora una variación de $b$ más genérica, del tipo:
$$
\Delta b = \begin{pmatrix}
10^{-p} \\
-10^{-p} \\
10^{-p} \\
-10^{-p}
\end{pmatrix} 
$$ 
calcula el menor $p$ que acarrea una pérdida de cifras significativas inferior a 2.

36. Usando Python:
  -  programa una función ```hilbert_matrix(n: int)``` que devuelva la matriz de Hilbert $H_n$ de dimensión $n$:
$$
H_{n}=\left(\begin{array}{ccccc}
1 & 2^{-1} & 3^{-1} & \ldots & n^{-1} \\
2^{-1} & 3^-1 & 4^{-1} & \ldots & (n+1)^{-1} \\
3^{-1} & 4^{-1} & 5^{-1} & \ldots & (n+2)^{-1} \\
\vdots      & \vdots      & \vdots      & \ddots & \vdots \\
n^{-1} & (n+1)^{-1} & (n+2)^{-1} & \ldots & (2n+1)^{-1}
\end{array}\right)
$$
es decir, el elemento en la posición $(i, j)$ de la matriz de Hilbert $H_n$ vale $(i+j+1)^{-1}$. 

  - Obtén el número de condición $\kappa_2 \left( H_n \right)$ para valores de $n$ entre 1 y 14, y grafica su comportamiento en función de $n$. 
  - El sistema $H_n x_n = b_n$ con $b_i = \sum_{j=1}^n \left( i+j+1 \right)^{-1}$ tiene solución exacta: $x_n = \left(1, 1, \overset{n}{\cdots}, 1\right)^T$. Analiza el error que se obtiene de resolver el sistema para $n$ crecientes.

37. La matriz de Vandermonde tiene la siguiente expresión genérica:
$$
V_{n}=\left(\begin{array}{ccccc}
1 & x_0 & x_0^2 & \cdots & x_0^n \\
1 & x_1 & x_1^2 & \cdots & x_1^n \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
1 & x_{n-1} & x_{n-1}^2 & \cdots & x_{n-1}^n
\end{array}\right)
$$
donde todos los $x_i$ son distintos. Calcula el numero de condición de $V_n$ para diversas elecciones de los $x_i$.

38. Usando Python, programa una función ```random_cond_info(n: int)``` que haga lo siguiente: partiendo de una matriz singular cualquiera de tamaño $n$ generada aleatoriamente, altera en una cierta cantidad uno de lo elementos de la matriz. Después, calcula el número de condición de esta nueva matriz, e imprime por pantalla un mensaje informando al usuario del número de cifras significativas en las que puede confiar al resolver un sistema de ecuaciones con dicha matriz como matriz de coeficientes.

39. Las matrices *sparse* tienen una gran importancia en computación, pues aparecen en un montón de campos representando matrices de conectividad, semejanza, activación, etc. Se caracterizan por estar compuestas en su mayoría por ceros, con algún elemento no nulo (minoritarios). Programa en Python una función ```sparse_matrix(n: int, sp: float) -> np.ndarray``` que devuelva una matriz cuadrada *sparse*, dado su tamaño ```n``` y su proporción de ceros frente al total de elementos ```sp``` (*sparsity*). Investiga si la *sparsity* tiene algún impacto en el número de condición de la matriz.  

40. *[Problema largo]* Una forma de la [ecuación de ondas](https://en.wikipedia.org/wiki/Wave_equation)
unidimensional puede escribirse como:

$$ \begin{aligned} \frac{\partial^{2}
u}{\partial t^{2}} &=c^{2} \frac{\partial^{2} u}{\partial x^{2}}, & x \in(0,
L), & t \in(0, T] \\
u(x, 0) &=I(x), & x & \in[0, L] \\
\frac{\partial}{\partial t} u(x, 0) &=0, & x & \in[0, L] \\
u(0, t) &=0, & t & \in(0, T] \\
u(L, t) &=0, & t & \in(0, T]
\end{aligned}
$$

donde $u$ es la velocidad, $t$ el tiempo, $x$ la coordenada espacial, $c$ la
velocidad de las ondas en el medio (constante), e $I(x)$ es la condición
inicial. Nótese que la cuerda estudiada parte del reposo ($\partial u(x, 0) /
\partial t  = 0$) y que sus extremos son fijos ($u(0, t) = u(L, t) = 0$).
 - Plantea una discretización espacial y temporal equiespaciada (con pasos $\Delta x$ y $\Delta t$, respectivamente),. Debe cumplirse que:
$$
CFL = c \frac{\Delta t}{\Delta x} \leq 0.4
$$
 - Discretiza la ecuación de ondas utilizando diferencias finitas centradas de
    orden dos, tanto en espacio como en tiempo. 
 - El esquema de diferencias finitas centradas de orden para el tiempo falla
    en el instante inicial. A partir del desarrollo de Taylor en $t + \Delta t$,
    obtén una expresión para el primer paso de tiempo. Ten en cuenta la
    condición inicial.
 - Implementa el algoritmo anterior. Comprueba los resultados frente a la siguiente solución analítica (conocida):
$$
u_e(x,t)= A \sin( \pi Lx) \cos ( \pi Lct)
$$
con $I(x)= A \sin(\pi Lx)$ (fija arbitrariamente el resto de parámetros). 